### PR TITLE
Allow configuring maximum word difference

### PR DESCRIPTION
**Pull request by Hlavtox**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?             | 9.0.x
| Description?      | This PR allows to set additional search settings. See more below.
| Type?             | improvement
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| How to test?      | Try to play with this field and see that it changes the amount of typos you can make in a word for the product to be found.
| UI Tests          | https://github.com/florine2623/testing_pr/actions/runs/11833815220 ✅ 
| Fixed issue or discussion?     | 
| Related PRs       | https://github.com/PrestaShop/autoupgrade/pull/1035
| Sponsor company   | TRENDO s.r.o.

### Context
- In Prestashop, there is fuzzy search enabled by default. This allows to find things even if there is a typo.
- If it doesn't find anything for a keyword, it will try to find list of simmilar words and takes the next closest one, until it finds something.
- It compares the words by a https://www.php.net/manual/en/function.levenshtein.php function, that compares how different two words are, and takes the closest word.

### The problem
- There is a limit to how the word can be different, by default it's 7. Which is unnecessarily high.
- So, imagine you are searching for something and there are cases where it should find nothing, but finds something irrelevant.
- For example, I search for `Blanco` and I can still find it with `Baaaaoaaa`.
- Simmilarily, this just invents completely irrelevant words.

### Solution
- Configure this to more sensible number, like 3 changes maximmum.
- So, I made a configuration field for this in search settings.

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `branch` should be `develop` or `8.2.x`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#branch))
* The `How to test` shouldn&#039;t be empty. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#how-to-test))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!

(Note: this is an automated message, but answering it will reach a real human)

**Comment by ps-jarvis** [comment]
This pull request seems to contain new translation strings. I have summarized them below to ease up review:

- `Admin.Shopparameters.Feature`
    - [ ] `Maximum acceptable word difference`
- `Admin.Shopparameters.Help`
    - [ ] `This option defines how much different can the alternative words found by fuzzy search be. Or, how many characters can be different/missing/added. The default value is 5.`

(Note: this is an automated message, but answering it will reach a real human)

**Comment by Hlavtox** [comment]
Ping @kpodemski @ShaiMagal @jf-viguier

**Review by FabienPapet — APPROVED** [review]

File: classes/Search.php
Comment (jf-viguier) [inline classes/Search.php:131]: this contant is no more in use

**Review by jf-viguier — APPROVED** [review]

File: classes/Search.php
Comment (Hlavtox) [reply classes/Search.php:131]: I left it there just for backward compatibility purpose, we can delete later. You know, if someone had overrides for example or just used it from some external module.

**Review by ShaiMagal — APPROVED** [review]

**Comment by gericfo** [comment]
Hello @Hlavtox
The “Needs autoupgrade PR” label has been automatically added to your Pull Request. Could you check if a PR is needed for the Autoupgrade? This is a gentle reminder as I'm manually reviewing all the PRs with this label! If it's not the case I will remove the label.
Thanks,
Géric

**Review by florine2623 — APPROVED** [review]
Hello @Hlavtox ,

Thanks for the PR !

Auto tests are green : https://github.com/florine2623/testing_pr/actions/runs/11833815220 ✅ 

I checked the BO, this option has been added :
Screenshot 2024-11-18 at 14 55 09

Works really good with the name of the product.

I tested with multiple strings. Tested with Title only ^^
For Title "brosse a cheveux", I can find the product with many word that are close to the title like "brasse", "brousse", "prosse", " brosser", "brozze", "cheval".
it works also when I search multiple words in the Search input area.

When I try to search a product with 5 character differences, I can't find the product.

https://github.com/user-attachments/assets/5bb7d675-2417-416f-bc9f-96783c6a5df6

It is QA ✅ !
Thanks!

**Comment by Hlavtox** [comment]
Thank you @florine2623!!!

**Comment by ps-jarvis** [comment]
PR merged, well done!

Message to @PrestaShop/committers: do not forget to milestone it before the merge.

**Pull request merged**
